### PR TITLE
Add CNN predictor

### PR DESCRIPTION
**Pull request by Milan933-coder**
## Describe your changes

Adds a regular-grid CNN predictor path for Neural-LAM.

This PR adds reusable node/grid tensor transforms, ResHRRR-inspired CNN layers, a `CNNPredictor` `StepPredictor`, CLI construction for `--model cnn_predictor`, checkpoint reconstruction support, and focused tests.

The CNN backbone is inspired by the ResHRRR model from HRRRCast: https://arxiv.org/abs/2507.05658.

It includes residual CNN blocks, Squeeze-and-Excitation, and optional FiLM conditioning, but it is not a full HRRRCast reproduction.

Multi-lead training, DDIM/diffusion sampling, and a U-Net backbone are left out of this PR.

No new runtime dependency is required.

## Issue Link

refs #590

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes. Not applicable for this initial predictor implementation.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by Milan933-coder** [comment]
@joeloskarsson please check it .I will move forward after it

**Comment by joeloskarsson** [comment]
Please add the full PR template.

Is the CNN the backbone from https://arxiv.org/abs/2507.05658? Would be good to add a reference. I would maybe add just a U-net as a first CNN architecture (to figure out the details that comes with requiring a regular grid), but I am not opposed to this one.

**Comment by Milan933-coder** [comment]
@joeloskarsson  Yes this is ResHRRR Inspired i used the exact blocks used in the paper but not the full implementation .
And about the unet Do you want me to add the setup where the user can change the cnn_backbone as unet or res_hrr .

**Comment by Milan933-coder** [comment]
@joeloskarsson yes  it is done .I have not addded the unet backbone .If you want i can add

**Review by Sir-Sloth-The-Lazy — COMMENTED** [review]
Sorry this was a mistake

File: neural_lam/cnn_layers.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+            padding=padding,
+            padding_mode=padding_mode,
+        )
+        self.act = nn.SiLU()
+        self.conv2 = nn.Conv2d(
+            channels,
+            channels,
+            kernel_size=kernel_size,
+            padding=padding,
+            padding_mode=padding_mode,
```
Comment (Sir-Sloth-The-Lazy) [inline neural_lam/cnn_layers.py:289]: `padding_mode` gets passed straight into `nn.Conv2d` here with nothing checking it against the actual grid shape. PyTorch's `reflect` mode needs `padding < input_size` on the padded axis, and `circular` needs the padding to not wrap around more than once. A grid with a spatial dimension of 1 will crash on the first forward pass here even with the default `cnn_kernel_size=3` and `reflect`, and `circular` will crash on bigger grids too once the kernel gets large enough. Since `--cnn_padding_mode` is just an open CLI choice right now, nothing stops someone from hitting this. Might be worth validating grid shape against kernel size and padding mode at construction time, or at least documenting the constraint.

File: neural_lam/models/step_predictors/cnn/cnn_predictor.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                prev_state,
+                prev_prev_state,
+                forcing,
+                self.expand_to_batch(self.grid_static_features, batch_size),
+            ),
+            dim=-1,
+        )
+        grid_features = node_to_grid(grid_features, self.grid_shape)
+
+        context = forcing.mean(dim=1) if self.cnn_film else None
```
Comment (Sir-Sloth-The-Lazy) [inline neural_lam/models/step_predictors/cnn/cnn_predictor.py]: This averages forcing over every grid node, so every cell ends up with the exact same FiLM conditioning no matter what the local forcing looks like (boundary vs interior, etc). Not a crash, just seems like it throws away the spatial signal a regional CNN would otherwise want to use per pixel. Was this a deliberate choice (matching how HRRRCast conditions globally), or was per node conditioning the intent?

File: neural_lam/models/step_predictors/cnn/cnn_predictor.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            persistent=False,
+        )
+        self.register_buffer(
+            "diff_std",
+            torch.tensor(
+                da_state_stats.state_diff_std_standardized.values,
+                dtype=torch.float32,
+            ),
+            persistent=False,
+        )
```
Comment (Sir-Sloth-The-Lazy) [inline neural_lam/models/step_predictors/cnn/cnn_predictor.py]: This block is basically identical to what's already in `BaseGraphModel.__init__` (`neural_lam/models/step_predictors/graph/base.py`, around L76 to L89). Since any delta predicting `StepPredictor` needs this rescale logic, seems like a good candidate to move up into the shared `StepPredictor.__init__` next to `state_mean`/`state_std` rather than copying it into every subclass.

**Review by Sir-Sloth-The-Lazy — COMMENTED** [review]
Hi, sorry for the empty review before, I this is really nice, just some issue that I found and thought might bring it to your attention. Thank for your time !

**Comment by Milan933-coder** [comment]
Okay I will look into it

**Comment by Milan933-coder** [comment]
@joeloskarsson i need yor review .This will be my first PR for the a issue who is being assigned to me so i am excited for your comments

**Comment by sadamov** [comment]
We had some discussions about additional model architectures/types at the last dev-meeting, that seem relevant here:

If we add new model types we always try to add the most general/basic variant first and not some complex paper implementation
New model types should preferrably be used in some experiment and not just added as a coding exercise to prevent bloat/dead code
We need a basic testsuite/notebook that evaluates a reproducible model forecast (e.g. based on DANRA); a certain minimum performance should be achieved.

see #722 for the proposed benchmark suite

**Comment by Milan933-coder** [comment]
okay Sir I will evaluate this method on DANRA dataset and will add a file that will also do the same thing

**Comment by sadamov** [comment]
Heads-up: we discussed the notebook format in the dev meeting and ran a poll in Slack afterwards, marimo won over Jupyter. Going forward notebooks in this repo are marimo `.py` files under `docs/notebooks/`, no `.ipynb` in git anymore.

Re the evaluation notebook I asked for above: that should be a marimo .py now, or just a plain pytest if that is simpler.

**Comment by Milan933-coder** [comment]
Sorry Sir ,I completely forgot busy with some schedule .I didn't get
e your latest commentt I think I have to write a evaluation notebook which I will do this week
